util/metrics.go

```diff
@@ -0,0 +1,140 @@
+		Namespace: metricsNamespace,
+		Name:      "in_flight_requests",
+		Help:      "A gauge of requests currently being served by the http server.",
+	})
+
+	httpRequestsTotal = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: metricsNamespace,
+			Name:      "http_requests_total",
+			Help:      "A counter for requests to the http server.",
+		},
+		[]string{"code", "method", "path"},
+	)
+
+	httpRequestDurationSeconds = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: metricsNamespace,
+			Name:      "http_request_duration_seconds",
+			Help:      "A histogram of latencies for requests to the http server.",
+		},
+		[]string{"code", "method", "path"},
+	)
+
+	httpRequestSizeBytes = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: metricsNamespace,
+			Name:      "http_request_size_bytes",
+			Help:      "A histogram of sizes of requests to the http server.",
+			Buckets: []float64{
+				16,
+				32,
+				64,
+				128,
+				256,
+				512,
+				1024,       /* 1KiB */
+				64 * 1024,  /* 64KiB */
+				256 * 1024, /* 256KiB */
+			},
+		},
+		[]string{"code", "method", "path"},
+	)
+
+	httpResponseSizeBytes = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: metricsNamespace,
+			Name:      "http_response_size_bytes",
+			Help:      "A histogram of response sizes for requests to the http server.",
+			Buckets: []float64{
+				16,
+				32,
+				64,
+				128,
+				256,
+				512,
+				1024,       /* 1KiB */
+				64 * 1024,  /* 64KiB */
+				256 * 1024, /* 256KiB */
+			},
+		},
+		[]string{"code", "method", "path"},
+	)
+)
+
+// MetricsMiddleware is a middleware used to measure every request received
+func MetricsMiddleware() mux.MiddlewareFunc {
+	// Rergister all metrics
+	prometheus.MustRegister(ServiceInfo)
+
+	prometheus.MustRegister(httpInFlightRequests)
+	prometheus.MustRegister(httpRequestsTotal)
+	prometheus.MustRegister(httpRequestDurationSeconds)
+	prometheus.MustRegister(httpRequestSizeBytes)
+	prometheus.MustRegister(httpResponseSizeBytes)
+
+	prometheus.MustRegister(NumberIndexedPackages)
+	prometheus.MustRegister(SearchProcessDurationSeconds)
+
+	return func(next http.Handler) http.Handler {
+		handler := next
+
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			route := mux.CurrentRoute(req)
+			path, _ := route.GetPathTemplate()
+			labels := prometheus.Labels{"path": path}
+
+			handler = promhttp.InstrumentHandlerCounter(httpRequestsTotal.MustCurryWith(labels), handler)
+			handler = promhttp.InstrumentHandlerDuration(httpRequestDurationSeconds.MustCurryWith(labels), handler)
+			handler = promhttp.InstrumentHandlerInFlight(httpInFlightRequests, handler)
+			handler = promhttp.InstrumentHandlerRequestSize(httpRequestSizeBytes.MustCurryWith(labels), handler)
+			handler = promhttp.InstrumentHandlerResponseSize(httpResponseSizeBytes.MustCurryWith(labels), handler)
+			handler.ServeHTTP(w, req)
+		})
+	}
+}
```

Comment:
I think it'd be good to move this to its own `metrics` package. For example when using `util.ServiceInfo` I wouldn't think this is anything related to monitoring.

Alternatively, use some `Metric` prefix/suffix for metrics-related variables, or "namespace" them by putting them in some kind of `MetricsManager` that contains everything related to metrics.